File: lib/tema/tokens.dart
/// Tokens de diseño: colores, espaciados, radios y tipografía estilo iOS.
///
/// Los colores son los semánticos de Cupertino. En Flutter son
/// `CupertinoDynamicColor`, es decir, **no valen tal cual**: hay que resolverlos
/// contra el contexto para que den el tono de claro o de oscuro. Ese es el error
/// clásico al usarlos, así que aquí se envuelven en la extensión [Paleta] y las
/// pantallas escriben `context.texto` en vez de tocar `CupertinoColors`.
library;

import 'package:flutter/cupertino.dart';

// ── Colores ──────────────────────────────────────────────────────────────────

/// Colores del tema, ya resueltos para el brillo actual.
extension Paleta on BuildContext {
  /// Fondo de pantalla.
  Color get fondo => CupertinoColors.systemGroupedBackground.resolveFrom(this);

  /// Fondo de las listas y tarjetas.
  Color get tarjeta =>
      CupertinoColors.secondarySystemGroupedBackground.resolveFrom(this);

  /// Fondo de la barra de navegación y la de pestañas.
  Color get barra => CupertinoColors.systemBackground.resolveFrom(this);

  Color get texto => CupertinoColors.label.resolveFrom(this);
  Color get textoSec => CupertinoColors.secondaryLabel.resolveFrom(this);
  Color get textoTer => CupertinoColors.tertiaryLabel.resolveFrom(this);
  Color get marcador => CupertinoColors.placeholderText.resolveFrom(this);

  Color get separador => CupertinoColors.separator.resolveFrom(this);

  /// Fondo de campos y píldoras.
  Color get relleno => CupertinoColors.tertiarySystemFill.resolveFrom(this);

  Color get acento => CupertinoColors.systemBlue.resolveFrom(this);
  Color get destructivo => CupertinoColors.systemRed.resolveFrom(this);
  Color get exito => CupertinoColors.systemGreen.resolveFrom(this);
}

/// Convierte un color de [coloresRutina] ('#0A84FF') en un [Color].
///
/// Es la única familia de colores literales de la app: identifica cada rutina en
/// el calendario y tiene que ser la misma en claro y en oscuro.
Color colorDesdeHex(String? hex, Color porDefecto) {
  if (hex == null || hex.length != 7 || !hex.startsWith('#')) return porDefecto;
  final valor = int.tryParse(hex.substring(1), radix: 16);
  return valor == null ? porDefecto : Color(0xFF000000 | valor);
}

// ── Espaciado ────────────────────────────────────────────────────────────────

const double xs = 4;
const double s = 8;
const double m = 12;
const double l = 16;
const double xl = 20;
const double xxl = 28;

// ── Radios ───────────────────────────────────────────────────────────────────

const double radioS = 8;
const double radioM = 10;
const double radioL = 14;
const double radioXl = 22;

// ── Tipografía (escala de iOS) ───────────────────────────────────────────────

const double largeTitle = 34;
const double title1 = 28;
const double title2 = 22;
const double title3 = 20;
const double headline = 17;
const double body = 17;
const double callout = 16;
const double subhead = 15;
const double footnote = 13;
const double caption = 12;

const FontWeight semibold = FontWeight.w600;
const FontWeight bold = FontWeight.w700;

/// Alto mínimo de una fila táctil, según las guías de Apple.
const double altoFila = 44;

File: app/theme/tokens.py
"""Tokens de diseño: colores, espaciados, radios y tipografía estilo iOS.

Los colores son los semánticos de Cupertino, que en Flutter ya son adaptativos:
el mismo token se resuelve a un valor distinto en tema claro y oscuro, así que
no hace falta duplicar la paleta ni escuchar cambios de tema a mano.
"""

import flet as ft

# ── Colores ──────────────────────────────────────────────────────────────────
FONDO = ft.CupertinoColors.SYSTEM_GROUPED_BACKGROUND          # fondo de pantalla
TARJETA = ft.CupertinoColors.SECONDARY_SYSTEM_GROUPED_BACKGROUND  # fondo de las listas
BARRA = ft.CupertinoColors.SYSTEM_BACKGROUND                   # appbar y tab bar

TEXTO = ft.CupertinoColors.LABEL
TEXTO_SEC = ft.CupertinoColors.SECONDARY_LABEL
TEXTO_TER = ft.CupertinoColors.TERTIARY_LABEL
PLACEHOLDER = ft.CupertinoColors.PLACEHOLDER_TEXT

SEPARADOR = ft.CupertinoColors.SEPARATOR
RELLENO = ft.CupertinoColors.TERTIARY_SYSTEM_FILL             # fondo de campos y píldoras

ACENTO = ft.CupertinoColors.SYSTEM_BLUE
DESTRUCTIVO = ft.CupertinoColors.SYSTEM_RED
EXITO = ft.CupertinoColors.SYSTEM_GREEN

# Paleta con la que se colorean las rutinas (mismo orden que COLORES_RUTINA
# en app/utils/conexionBD.py).
PALETA_RUTINAS = ["#0A84FF", "#30D158", "#FF9F0A", "#FF375F",
                  "#BF5AF2", "#40C8E0", "#5E5CE6", "#FFD60A"]

# ── Espaciado ────────────────────────────────────────────────────────────────
XS = 4
S = 8
M = 12
L = 16
XL = 20
XXL = 28

# ── Radios ───────────────────────────────────────────────────────────────────
RADIO_S = 8
RADIO_M = 10
RADIO_L = 14
RADIO_XL = 22

# ── Tipografía (escala de iOS) ───────────────────────────────────────────────
LARGE_TITLE = 34
TITLE1 = 28
TITLE2 = 22
TITLE3 = 20
HEADLINE = 17
BODY = 17
CALLOUT = 16
SUBHEAD = 15
FOOTNOTE = 13
CAPTION = 12

SEMIBOLD = ft.FontWeight.W_600
BOLD = ft.FontWeight.W_700

# Alto mínimo de una fila táctil, según las guías de Apple.
ALTO_FILA = 44

# ── Temas de la página ───────────────────────────────────────────────────────

def _transiciones():
    """Transiciones de empuje estilo iOS en todas las plataformas."""
    return ft.PageTransitionsTheme(
        ios=ft.PageTransitionTheme.CUPERTINO,
        android=ft.PageTransitionTheme.CUPERTINO,
        macos=ft.PageTransitionTheme.CUPERTINO,
        windows=ft.PageTransitionTheme.CUPERTINO,
        linux=ft.PageTransitionTheme.CUPERTINO,
    )


def tema_claro():
    return ft.Theme(page_transitions=_transiciones(), font_family=None, use_material3=True)


def tema_oscuro():
    return ft.Theme(page_transitions=_transiciones(), font_family=None, use_material3=True)
